README.md

```diff
@@ -112,6 +112,21 @@ If `step` of instances of type `T` are known at compile time, return that step.
 Otherwise, returns `nothing`. For example, `known_step(UnitRange{Int})` returns
 `one(Int)`.
 
+## is_cpu_column_major(::Type{T})
```

Comment:
why cpu?

Reply:
Do we also want to consider accommodating other memory layouts that have been used in improving performance of array operations [like these](https://github.com/JuliaMatrices/ArrayLayouts.jl/blob/master/src/memorylayout.jl#L54)?

Reply:
@ChrisRackauckas 
"cpu" because I wanted it to be `false` for GPUArrays.
Given that [AbstractDeviceArrays <: DenseArray](https://github.com/JuliaGPU/GPUArrays.jl/blob/b34a724be5b4596ac74636adc01fac6768aa30e4/src/device/abstractarray.jl#L15) to opt into the `StrideArray` interface (even though `LinearAlgebra` routines presumably don't work), I wanted to make it clear they shouldn't opt into this.

@Tokazama 
The `stridelayout` function can efficiently handle row-major or increasing vs decreasing strides, e.g.:
```julia
julia> using ArrayInterface: stridelayout

julia> is_row_major(x) = stridelayout(x) === (2,1,(2,1))
 is_row_major (generic function with 1 method)

julia> B = rand(3,4);

julia> @code_typed is_row_major(B')
 CodeInfo(
1 ─     return true
) => Bool

julia> @code_typed is_row_major(B)
 CodeInfo(
1 ─     return false
) => Bool
```
But perhaps I should add density. Knowing that certain dims are dense could potentially allow subsets of loops to undergo `CartesianIndexing` -> `LinearIndexing` style loop fusion, even if on the whole the `IndexStyle` is `IndexCartesian()`.

Reply:
Yeah, I was specifically thinking of dense arrays but didn't want to chance overlooking something else. Using IndexStyle to accomplish this is pretty clever.

Reply:
I added density info on a per-stride basis.
```julia
julia> A = rand(3,4,5);

julia> stridelayout(PermutedDimsArray(A,(3,1,2)))
 (2, 1, (3, 1, 2), (true, true, true))

julia> stridelayout(@view(PermutedDimsArray(A,(3,1,2))[2,1:2,:]))
 (1, 1, (1, 2), (true, false))

julia> stridelayout(@view(PermutedDimsArray(A,(3,1,2))[2:3,1:2,:]))
 (2, 1, (3, 1, 2), (false, false, true))
```
Unfortunately, we can't tell from type info whether a slice of an array was partial or complete:
```julia
julia> s2 = Base.Slice(Base.OneTo(2))
 Base.Slice(Base.OneTo(2))

julia> view(A,s2,s2,s2)
 2×2×2 view(::Array{Float64,3}, :, :, :) with eltype Float64:
[:, :, 1] =
 0.329658  0.543774
 0.350255  0.0817058

[:, :, 2] =
 0.310211  0.126501
 0.587191  0.884039
```
as `Base.Slice{Base.OneTo{Int64}}` is the type of a `:` slice. This means we have to be very conservative with base `SubArray`s, but packages (like PaddedMatrices.jl) might want to provide more compile time info for their types.

The finer granularity would allow for the partial linear-vs-cartesian indexing.
Other reasons to have this information:
- using linear indexing to decide what to do for something like a `vec` implementation is a little goofy.
- LoopVectorization often spills scalar integer registers (i.e., `r*` registers). Many of these registers are used for holding array strides, and integer multiples of these strides. E.g., with a AVX512 column-major matmul kernel for `C = A * B`, it wants to hold `stride(A,2), 3stride(A,2), 5stride(A,2), stride(C,2), 3stride(C,2), 5stride(C,2)`. So if we know that `stride(A,2) == stride(C,2)`, we could free 3 registers, save a couple calculations, and a bunch of integer load/stores from the stack when we get less spills. I'm not sure how to implement this optimization / let a user guarantee that they aren't writing 100% valid code like
```julia
A = rand(4,5);
C = rand(40,50);
for m in axes(A,1), n in axes(A,2)
    C[m,n] = 2A[m,n]
end
```
So knowing that `A` and `C` are both dense still isn't enough to guarantee that the strides are equal.

Not particularly relevant to this issue, but I wonder if there'd be a  good way to communicate size equality. Users will often check for equal sizes in front of a loop, and `eachindex` will guarantee this as well
```julia
for I in eachindex(A,C) # checks for equivalent size
   
end
```
Because `eachindex` isn't always appropriate (e.g., when you don't have exact correspondence between axes of the arrays), I think it'd be a good to have something like an `eachindexalongaxis`:
```julia
function mymatmulkernel!(C, A, B)
    @avx for m in eachindexalongaxis((C,A),(1,1)), n in eachindexalongaxis((C,B),(2,2)), k in eachindexalongaxis((A,B),(2,1))
        C[m,n] += A[m,k] * B[k,n]
    end
end
```
where it checks sizes. `eachindexalongaxis` or `each_index_along_axis` doesn't really roll off the tongue, and I don't know where such a function should live.
I'd also have to think a little more about how to use it and actually pass the information of dimension equality (to be combined with density guarantees) to deduce stride-equality. The hacky way would be to just pattern match `eachindex` and `eachindexalongindex` in the macro, handling dimension equality through the expression, while handling density info based on types.

Reply:
I'll make a separate issue about the axes specific indexing.

Reply:
The only think I have left to say about this one is that you may want to include a some explanation or a link so people understand the "cpu" part. It's obviously not a huge deal, but I anticipate it becoming a common question that becomes irksome to continually explain.

Reply:
How about a comment simply saying that it should return `false` for any sort of GPUArray?

Reply:
That would work.